subsys/net/lib/sockets/sockets_misc.c

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2019 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <net/socket.h>
+
+#define DEFAULT_HOSTNAME "zephyr-dflt"
```

Comment:
We have CONFIG_NET_HOSTNAME that should contain the default value. If it is not defined, then a default value specified here could be used instead. Also the `zephyr-dflt` sounds weird, why not call it `zephyr` instead?

Reply:
> If it is not defined, then a default value specified here could be used instead.

But that's exactly how it works!

> Also the `zephyr-dflt` sounds weird, why not call it `zephyr` instead?

It was exactly made weird (and distinguishable, so easy to find in a sense of few/now false positives), so people could come to the need of using CONFIG_NET_HOSTNAME_* to make it non-weird.

Reply:
> But that's exactly how it works!

To clarify, yes, I started with `#ifdef CONFIG_NET_HOSTNAME_ENABLE`, but figured I exactly can simplify by checking for NULL which net_hostname_get() is statically defined to if that option is not enabled.

Reply:
Proves that meta-level `#ifdef`s rule and easy for cognition, while all those flat `if`s suck ;-). (But `#ifdef`s must be indented for real bliss!)

Reply:
Why not making net_get_hostname() always returning non null instead?

At net_hostname_init() it could check if CONFIG_NET_HOSTNAME is relevant (for instance if it's "" then it would need to generate one).

That would avoid handling the case here.

Reply:
> Why not making net_get_hostname() always returning non null instead?

Why? For small systems, hostname is strictly optional, and why we have it disabled by default. But for POSIX systems, it's expected to exist, that's why there's a default here.

Reply:
Yes but now that we are switching to socket API as the unique API for users, we should probably make hostname always built. And the code isn't really heavy so that should be fine for small systems too

Reply:
Actually, instead of building the hostame.c, in include/net/hostname.h have the inline version of net_get_hostname returning a default value.

Reply:
> Yes but now that we are switching to socket API as the unique API for users

Well, "sockets API" itself consists of different sub-APIs. There's data communication API per se (socket() and friends), which we do make the default in Zephyr (without disallowing those in need to use native APIs if they suit there usecase). Then there's "name resolution" API, which already goes towards "optional" (that API is pretty bloated). And then there's different support funcs, like this gethostname(). They're more part of POSIX subsys, then of bare Socket API. Native Zephyr-targetting apps are not encouraged to use them at all. They're there to port existing POSIX apps to Zephyr (and why I add it here - I port such an app).

And I'm telling that as the same guy who told recently "let's keep different socket types implementations separate for now, we'll refactor common code once they stabilize". For cases where it's clear what's the optimal solution, I don't call to add a single byte of code than necessary, Zephyr needs to start pay real attention (example: recently, MicroPython Zephyr port stopped to fit into Arduino 101, and seriously doubt that happened on MicroPython side).

Reply:
> Zephyr needs to start pay real attention 

I agree, actually once legacy net_pkt will be removed net stack will have shrunk quite a lot (having both currently makes it heavy for the time being obviously). However I am not entirely sure other subsystems care about small systems, unfortunately.

Reply:
To me it sounds like that if CONFIG_NET_HOSTNAME_ENABLE is not set, then calling gethostname() should return -1 and set errno to ENOTSUP. Then we do not need to handle the "hostname not set" case at all here.

Reply:
> To me it sounds like that if CONFIG_NET_HOSTNAME_ENABLE is not set, then calling gethostname() should return -1 and set errno to ENOTSUP.

It will immediately break the application I'm porting, and likely a lot more other applications. And it will be hard to link that failure to the need to set a specific option in Zephyr config.

Really guys, there's little to bikeshed about. Zephyr native apps very rarely need a hostname. POSIX apps assumes the hostname is there. This patch is the simple and obvious way to achieve that. (I agree that it probably makes sense to make a default hostname here "zephyr-host", still distinguishable, but more clear. Will do.)

Reply:
Imo, simplest way is: 
"in include/net/hostname.h have the inline version of net_get_hostname returning a default value."

Reply:
> Imo, simplest way is:
"in include/net/hostname.h have the inline version of net_get_hostname returning a default value."

Please understand my view: this doesn't bring anything to to the table. If CONFIG_NET_HOSTNAME_ENABLE is set, it works almost exactly like you say: there's a static array of the exact size of the hostname (with a default, adjustable by user in config). But CONFIG_NET_HOSTNAME_ENABLE is not set, we save those handful, but still bytes, which aren't needed for 90% of native applications.

Reply:
> But CONFIG_NET_HOSTNAME_ENABLE is not set, we save those handful, but still bytes, which aren't needed for 90% of native applications.

As long as net_hostname_get() is not used, it won't end up in the image. And if it is used, it will be exactly as you did with your own default.
Check how is net_hostname_get() being used, but I am pretty sure it it at I said. (for instance net shell do not call the function if CONFIG_NET_HOSTNAME_ENABLE  is not set).

Either way, note that I am not -1 your pr. I am just trying to avoid having default hostname handling in various places.

Reply:
I agree with @tbursztyka here, why not change the include/net/hostname.h to return default hostname if hostname support is not enabled. Then we do not need to add default hostname to socket code.

Reply:
> I agree with @tbursztyka here, why not change the include/net/hostname.h to return default hostname if hostname support is not enabled.

You see guys, this comes from much larger set of changes. I thought this one is very obvious and clear, and pushed it ahead of the rest, to avoid 30-commits reviews later.

Now you bikeshed it into changing things which don't really need to be changed to just implement that feature. So, what exactly do you want me to do? Remove CONFIG_NET_HOSTNAME_ENABLE completely? Because if you look at that code, it totally makes no sense to have one hardcoded default, another user-configurable default, option to select between them - that's quite a mess.

Also, saying that if net_hostname_get() isn't used, it simply won't get into the image (with extra default, not used in 90% of cases) isn't exactly true in the presence of userspace. (Which is heck, isn't exactly supported for these funcs, which is a separate task. And another reason I don't want to make too many changes now, to rework them again later for userspace.)

Reply:
> Because if you look at that code, it totally makes no sense to have one hardcoded default, another user-configurable default, option to select between them - that's quite a mess.

Exactly, that is why we are proposing that the hardcoded default is only added to one place, to hostname.h. Then in socket code there is no need to do anything special, just call the net_hostname_get() and you will get either the default one (regardless of if user has enabled hostname support or not) or the user configured hostname.

Reply:
I think we can keep it, at least for now.